code/modules/ninja/energy_katana.dm

```diff
@@ -1,30 +1,87 @@
-/obj/item/weapon/katana/energy
+/obj/item/weapon/dash
+	name = "energy katana"
+	var/max_charges = 3
+	var/current_charges = 3
+	var/charge_rate = 30 //In deciseconds
+	var/dash_toggled = FALSE
+
+/obj/item/weapon/dash/proc/charge()
+	current_charges = Clamp(current_charges + 1, 0, max_charges)
+	if(istype(loc, /mob/living))
+		to_chat(loc, "<span class='notice'>[src] now has [current_charges]/[max_charges] charges.</span>")
+
+/obj/item/weapon/dash/attack_self(mob/user)
+	dash_toggled = !dash_toggled
+	if(dash_toggled)
+		to_chat(loc, "<span class='notice'>[src] is now in dash mode.</span>")
```

Comment:
`user` not `loc`.
Also this can just be two lines with
`to_chat(user, "<span class='notice'>You [dash_toggled ? "enabled" : "disabled"] the dash function on [src]</span>")`